Also propagate storage_access_should_throw_ when copying tensor metadata

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54079 Add Tensor::is_cpu, genericize TensorIterator
* #54034 Refactor tensor_new.cpp to use TensorOptions instead of DispatchKey
* #54016 Delete denseTypeIdWithDefault and toDense
* #54013 Delete defunct ComplexCPU/ComplexCUDA dispatch keys
* #53972 Make storage access error NotImplementedError
* #53823 Compute type_equal() without reference to backend()
* **#53816 Also propagate storage_access_should_throw_ when copying tensor metadata**

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27036574](https://our.internmc.facebook.com/intern/diff/D27036574)

ezyang committed:
Also propagate storage_access_should_throw_ when copying tensor metadata

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Also propagate storage_access_should_throw_ when copying tensor metadata"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

swolchok commented:
What if you copy_tensor_metadata from a TensorImpl whose storage access should throw(say, a SparseTensorImpl) into one where it shouldn't (say, regular old TensorImpl)? The TensorImpl public interface permits this.

ezyang commented:
@swolchok I suspect there's an implicit contract saying that the subclass has to be the same in the source and the destination. Would have to add an assert to this effect to check though.

ezyang committed:
Update on "Also propagate storage_access_should_throw_ when copying tensor metadata"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Also propagate storage_access_should_throw_ when copying tensor metadata"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27036574](https://our.internmc.facebook.com/intern/diff/D27036574)

[ghstack-poisoned]

ezyang committed:
Update on "Also propagate storage_access_should_throw_ when copying tensor metadata"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27036574](https://our.internmc.facebook.com/intern/diff/D27036574)

[ghstack-poisoned]

ezyang committed:
Update on "Also propagate storage_access_should_throw_ when copying tensor metadata"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27036574](https://our.internmc.facebook.com/intern/diff/D27036574)

[ghstack-poisoned]

ezyang committed:
Update on "Also propagate storage_access_should_throw_ when copying tensor metadata"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27036574](https://our.internmc.facebook.com/intern/diff/D27036574)

[ghstack-poisoned]

ezyang committed:
Update on "Also propagate storage_access_should_throw_ when copying tensor metadata"


Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27036574](https://our.internmc.facebook.com/intern/diff/D27036574)

[ghstack-poisoned]

albanD reviewed (approved):
lgtm
